### DISCUSSION TITLE
`macos_icons_hack()` can be removed and replaced with a little locale configuration guide

### DISCUSSION THREAD

**Discussion by acc15** [jarun/nnn#1899] [Ideas]
Hello guys, 

Stuck on this bug:

![nnn](https://github.com/jarun/nnn/assets/4038899/5a9ed1e1-4642-462b-a399-cb494f9f89c0)

~ 1 year ago MacOS icon hack was added here https://github.com/jarun/nnn/pull/1711
But as you see it fixes only half of the problem (selections are still broken). 

I've found a proper way how to solve all this nasty MacOS icon problems without using ANY dirty hacks.

### Story

Some time ago Nerd icons Material Design icon set was moved to `0xf0000`-`0xf1af0` character range  (https://github.com/ryanoasis/nerd-fonts/wiki/Glyph-Sets-and-Code-Points)

This completely screws ncurses and strange things starts to happen:
 
* https://github.com/jarun/nnn/issues/1692 
* https://github.com/jarun/nnn/issues/1802
* ...and many others icons related issues on Mac.

Locale definitions are located in `/usr/share/locale` dir. 
Depending on user configuration some of this predefined locales is selected, and all locales ending with `.UTF-8` has symlink for `LC_CTYPE` pointing to `../UTF-8/LC_CTYPE`. 
So character widths are baked into binary locale definition file: `/usr/share/locale/UTF-8/LC_CTYPE`.
`UTF-8/LC_CTYPE` file shipped with MacOS makes all characters in `0xf0000`-`0xf000e` range to be 2 character wide. 
Which is a root cause of all those MacOS icon issues.

I've checked `ncurses` code and he is relies solely on [`wcwidth`](https://www.man7.org/linux/man-pages/man3/wcwidth.3.html) calls (not `wcswidth`) to get character widths. `wcwidth` itself relies on `LC_CTYPE` locale category. 

So to fix icon problems we need to create a proper `LC_CTYPE` file.

### Locale building

To build correct UTF-8 `LC_CTYPE` on `*BSD` systems there is a tool called [`mklocale`](https://man.openbsd.org/mklocale.1)
And luckily OpenBSD project shares pretty good UTF-8 `LC_CTYPE` source file: https://github.com/openbsd/src/blob/master/share/locale/ctype/en_US.UTF-8.src

MacOS requires only a single modification to this file - `ENCODING        "UTF8"` -> `ENCODING        "UTF-8"` (additional dash).

After that this file can be converted to binary form using:

    mklocale -o LC_CTYPE en_US.UTF-8.src  

### Installation
 
There are many ways about how to install it. 
According to [`setlocale`](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/setlocale.3.html) Apple libc will look up next directories (in this order) for locale definition files:

     $PATH_LOCALE/locale/category
     /usr/share/locale/locale/category

Modification of `/usr/share/locale` requires SIP disabling (https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection) and probably might break some system update in future, so its a bad way and generally should be avoided. 

So to install:

1. `sudo mkdir -p /usr/local/share/locale/UTF-8`
1. `sudo cp <path to binary LC_TYPE> /usr/local/share/locale/UTF-8`
1. in `~/.zshrc` add `export PATH_LOCALE=/usr/local/share/locale` and `export LC_CTYPE=UTF-8`

After this steps icons in `nnn` will just work without any hacks.

So i think `macos_icon_hack()` can be completely removed and replaced with this little guide about how to make a proper locale on MacOS. 

What do u think?

**Comment by N-R-K** [comment]
Nice detailed analysis.

> So i think macos_icon_hack() can be completely removed and replaced with this little guide about how to make a proper locale on MacOS.

This seems like a huge leap through.

a. This requires the user to go through some pretty convoluted steps to get icons working
b. The way you're exporting those vars in `.zshrc` will make this change affect other applications too, not just `nnn`.

Problem `b` can probably be solved by making a wrapper for `nnn` which sets up those env vars without exporting it for everything else. But problem `a` still remains, though.

**Reply by acc15** [reply]
a. What about script? https://github.com/acc15/macos_openbsd_utf8_locale/blob/master/macos_openbsd_utf8_locale.zsh
b. Yes, it will fix any other app which uses nerd material icons. Another options is an alias or wrapper script. But there is already - quitcd.

**Reply by N-R-K** [reply]
> Yes, it will fix any other app which uses nerd material icons.

Shouldn't this issue (and workaround) be submitted to nerd icons repo then?

**Reply by pa-0** [reply]
>Yes, it will fix any other app which uses nerd material icons

So does this mean a user actually stands to benefit from exporting the variables globally as opposed to using them in `nnn` exclusively?
